third_party/amd/lib/TritonAMDGPUTransforms/CanonicalizePointers.cpp

```diff
@@ -650,6 +814,174 @@ class ConvertAddPtrOp : public PointerCanonicalizationPattern<tt::AddPtrOp> {
     if (propagateAtrs)
       fatPtrs[nextFatPtr].attributes =
           fatPtrs.at({fatPtrBase, fatPtrOffset}).attributes;
+    return success();
+  }
+
+private:
+  LogicalResult
+  rewriteSmallTensorPtr(tt::AddPtrOp addPtrOp, OneToNOpAdaptor adaptor,
+                        ConversionPatternRewriter &rewriter) const {
+    ValueRange remappedPtr = adaptor.getPtr();
+    ValueRange nonRemappedOffset = adaptor.getOffset();
+
+    assert(remappedPtr.size() == 2 && nonRemappedOffset.size() == 1 &&
+           "expected to be satisified in caller");
+
+    Value fatPtrBase = remappedPtr[0];
+    Value fatPtrOffset = remappedPtr[1];
+    Value origOffset = nonRemappedOffset[0];
+    Location curLoc = addPtrOp.getLoc();
+
+    RewriterBase::InsertionGuard guard(rewriter);
+    rewriter.setInsertionPoint(addPtrOp);
+
+    const auto &oldAttr = fatPtrs.at({fatPtrBase, fatPtrOffset});
+
+    LDBG("smal-tensor addPtr: " << addPtrOp);
+    LDBG("   - with tensor-base: " << oldAttr.smallTensorBase);
+    LDBG("   - with originl offset: " << origOffset);
+    LDBG("   - fatPtr base: " << fatPtrBase);
+    LDBG("   - fatPtr offst: " << fatPtrOffset);
+
+    // This loop goes over all offset expressions and try to decompose them
+    // into uniform and non-uniform parts, and accumulte these two parts
+    // respectively.
+    //
+    // Each iteration decompose the given offset expression into 3 categories
+    //  - uniform value
+    //  - non-uniform value
+    //  - const-tensors value, i.e. a tensor whose elements are equal.
+
+    Value offsetExprs[] = {fatPtrOffset, origOffset};
+
+    SmallVector<Value> uniforms;
+    SmallVector<Value> nonUniforms;
+    SmallVector<std::pair</*tensor*/ Value, /*element*/ Value>> constTensors;
+
+    ScalarToSplatMap scalarToSplatMap;
+
+    for (auto offsetIter : offsetExprs) {
+      Type offsetType = offsetIter.getType();
+      if (offsetType.isIntOrIndexOrFloat()) {
```

Comment:
Offset cannot be float so just `isIntOrIndex()` so you can drop the following `assert`?